# NLP Task 6. Dialog Agent

> **Выполнил**: Давидчук Никита Олегович <br>
> **Датасет**: `dialog_talk_agent.xlsx`

### Немного о методологии работы

Стоит разделить работу на две части:
1. Подготовка данных
2. Решение задачи ответа на вопросы

В случае разбора данных, по первому осмотру файла можно сразу сказать, что первоначально стоит избавиться от пропусков

![Пропуски... повсюду!](https://i.imgur.com/TBdjK17.png)

Здесь есть два вида пропусков: 
* Пропуски в столбце `Text Response`
* Пропуски в рядах (полностью пустые ряды `NaN`, `NaN`)

И в первом, и во втором случае, исправить это с помощью `pandas` достаточно просто.
    
```python
# Удаление пропусков в столбце Text Response
df.dropna(how='all', inplace=True)

# Удаление пропусков в рядах
df = df['Text Response'].ffill()
```

После этого можно приступить к анализу данных. Для этого можно построить график распределения длин ответов.
(Но делать этого я не стал 😇)

Есть лишь одно допущение -- сохранение данных в формате `.csv` и разделителем `;` (вместо `,`), чтобы не было проблем с предложениями, которые имеют запятые.

### Импортируем библиотеки

In [1]:
import pandas as pd

Для корректной работы `pandas` с .xlsx файлами, необходимо установить `openpyxl`.

In [ ]:
%pip install openpyxl -q

Note: you may need to restart the kernel to use updated packages.


Прочтем файл и посмотрим на данные

In [14]:
# open the dialog_talk_agent.xlsx
df = pd.read_excel('dialog_talk_agent.xlsx')

In [15]:
df.head()

Context  \
0  Tell me about your personality   
1       I want to know you better   
2                 Define yourself   
3               Describe yourself   
4          tell me about yourself   

                                   Text Response  
0    Just think of me as the ace up your sleeve.  
1  I can help you work smarter instead of harder  
2                                            NaN  
3                                            NaN  
4                                            NaN

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1592 entries, 0 to 1591
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Context        1507 non-null   object
 1   Text Response  233 non-null    object
dtypes: object(2)
memory usage: 25.0+ KB


Удалим все пропуски, где весь ряд состоит из `NaN`

In [18]:
# drop all rows if all columns are NaN
df.dropna(how='all', inplace=True)

Заполним все пустые ячейки в `Text Response` от первого входа до последующего предыдущими значениями

In [17]:
# expand first entry in 'Text Response' column to multiple rows before the next non-NaN entry
df['Text Response'] = df['Text Response'].ffill()

### Немного интересного и мемного usecase

В некоторых значениях колонок `Text Response` и `Context` существуют значения, которые неоднозначно можно интепритировать как булевы
(`True`, `False`)

Изменим тип всех значений в таблице на строковые и приведем их к нижнему регистру

In [19]:
# convert all bool values to string values (True -> 'True', False -> 'False') to lower case
df = df.astype(str).apply(lambda x: x.str.lower())

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1592 entries, 0 to 1591
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Context        1592 non-null   object
 1   Text Response  1592 non-null   object
dtypes: object(2)
memory usage: 25.0+ KB


In [21]:
df.head()

Context  \
0  tell me about your personality   
1       i want to know you better   
2                 define yourself   
3               describe yourself   
4          tell me about yourself   

                                   Text Response  
0    just think of me as the ace up your sleeve.  
1  i can help you work smarter instead of harder  
2  i can help you work smarter instead of harder  
3  i can help you work smarter instead of harder  
4  i can help you work smarter instead of harder

In [22]:
df[:20]

Context  \
0   tell me about your personality   
1        i want to know you better   
2                  define yourself   
3                describe yourself   
4           tell me about yourself   
5                    all about you   
6     tell me some stuff about you   
7        talk some stuff about you   
8              talk about yourself   
9                   about yourself   
10                     who are you   
11              introduce yourself   
12   i want to know more about you   
13                    what are you   
14        what is your personality   
15                   say about you   
16               tell me about you   
17                why are you here   
18                             nan   
19            are you 21 years old   

                                        Text Response  
0         just think of me as the ace up your sleeve.  
1       i can help you work smarter instead of harder  
2       i can help you work smarter instead of harder  
3       i can help you work smarter instead of harder  
4       i can help you work smarter instead of harder  
5       i can help you work smarter instead of harder  
6       i can help you work smarter instead of harder  
7       i can help you work smarter instead of harder  
8       i can help you work smarter instead of harder  
9       i can help you work smarter instead of harder  
10      i can help you work smarter instead of harder  
11      i can help you work smarter instead of harder  
12      i can help you work smarter instead of harder  
13      i can help you work smarter instead of harder  
14      i can help you work smarter instead of harder  
15      i can help you work smarter instead of harder  
16      i can help you work smarter instead of harder  
17      i can help you work smarter instead of harder  
18      i can help you work smarter instead of harder  
19  i'm a relatively new bot, but i'm wise beyond ...

Сохраним данные в формате `.csv` и разделителем `;`

In [26]:
# save the dataframe to a new excel file
df.to_excel('dialog_talk_agent_clean.xlsx', index=False)

In [27]:
# convert to csv, but dont use commas as delimiters
df.to_csv('dialog_talk_agent_clean.csv', index=False, sep=';')

Убедимся в корректности сохранения и чтения

In [28]:
pd.read_csv('dialog_talk_agent_clean.csv', sep=';').info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1592 entries, 0 to 1591
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Context        1507 non-null   object
 1   Text Response  1592 non-null   object
dtypes: object(2)
memory usage: 25.0+ KB


## Создание модели на основе DialoGPT

Обучим модель на основе DialoGPT на наших данных. 
Microsoft DialoGPT представляет собой модель, обученную на диалогах из датасета [DailyDialog](https://arxiv.org/abs/1710.03957). Основной язык обучения - английский. Мы берем предобученную модель за основу и пытаемся произвести `fine-tuning` на наших данных. <br>

Ниже публичный код блокнота, который расположен в примерах сообщества Hugging Face 🤗. Код подлежит рефакторингу, но в рамках общей работы достаточно отличного запуска. <br>

Отличный туториал по работе с блокнотом расположен здесь: 
> [DialoGPT Espaniola tutorial](https://nathancooper.io/i-am-a-nerd/chatbot/deep-learning/gpt2/2020/05/12/chatbot-part-1.html) from Nathan Cooper.

GPT2 - это то, что называется моделью авторегрессивного языка. Это может показаться сложным, но на самом деле это тихо и просто, так что давайте разберемся, что это значит. Авторегрессивный означает, что вывод модели возвращается в модель в качестве входных данных. Вот хороший пример того, как это работает:

![transformers architecture](https://github.com/ncoop57/i-am-a-nerd/blob/master/images/autoregressive.gif?raw=1)

Задачу ответа бота на вопрос пользователя без *косинусного сходства (расстояния)* можно решить при помощи модели `Sentence Similarity` (sentence transformer model type). Но в данном случае мы будем использовать модель `DialoGPT` для генерации ответа на вопрос пользователя. <br>

> Sentence Similarity: https://huggingface.co/tasks/sentence-similarity

### Импортируем модели из Hugging Face 🤗 и загружаем предобученную модель DialoGPT

In [ ]:
from transformers import AutoModelWithLMHead, AutoTokenizer
import torch

tokenizer = AutoTokenizer.from_pretrained("microsoft/DialoGPT-small")
model = AutoModelWithLMHead.from_pretrained("microsoft/DialoGPT-small")

In [3]:
"""
Fine-tuning the library models for language modeling on a text file (GPT, GPT-2, BERT, RoBERTa).
GPT and GPT-2 are fine-tuned using a causal language modeling (CLM) loss while BERT and RoBERTa are fine-tuned
using a masked language modeling (MLM) loss.
"""

import glob
import logging
import os
import pickle
import random
import re
import shutil
from typing import Dict, List, Tuple

import pandas as pd
import numpy as np
import torch

from sklearn.model_selection import train_test_split

from torch.nn.utils.rnn import pad_sequence
from torch.utils.data import DataLoader, Dataset, RandomSampler, SequentialSampler
from torch.utils.data.distributed import DistributedSampler
from tqdm.notebook import tqdm, trange

from pathlib import Path

from transformers import (
    MODEL_WITH_LM_HEAD_MAPPING,
    WEIGHTS_NAME,
    AdamW,
    AutoConfig,
    AutoModelWithLMHead,
    AutoTokenizer,
    PreTrainedModel,
    PreTrainedTokenizer,
    get_linear_schedule_with_warmup,
)


try:
    from torch.utils.tensorboard import SummaryWriter
except ImportError:
    from tensorboardX import SummaryWriter

# Configs
logger = logging.getLogger(__name__)

MODEL_CONFIG_CLASSES = list(MODEL_WITH_LM_HEAD_MAPPING.keys())
MODEL_TYPES = tuple(conf.model_type for conf in MODEL_CONFIG_CLASSES)

Создадим аргументы для трейнера модели:

* `model_name_or_path` - имя модели или путь к ней
* `output_dir` - путь к папке, в которой будет сохранена модель
* `overwrite_output_dir` - перезаписать папку с моделью, если она уже существует
* `num_train_epochs` - количество эпох обучения
* `per_device_train_batch_size` - размер батча
* `per_device_eval_batch_size` - размер батча
* `save_steps` - количество шагов, после которых будет сохранена модель
* `save_total_limit` - количество моделей, которые будут сохранены
* `evaluation_strategy` - стратегия оценки модели
* `logging_steps` - количество шагов, после которых будет сохранен лог
* `logging_dir` - путь к папке, в которой будет сохранен лог
* `block_size` - размер блока токенайзера
* `warmup_steps` - количество шагов, после которых будет начато обучение
* `learning_rate` - скорость обучения
* `device` - устройство, на котором будет обучаться модель (в моем случае это Apple Silicon M1 GPU (`mps`))

In [4]:
# Args to allow for easy convertion of python script to notebook
class Args():
    def __init__(self):
        self.output_dir = 'new_output-small'
        self.model_type = 'gpt2'
        self.model_name_or_path = 'microsoft/DialoGPT-small'
        self.config_name = 'microsoft/DialoGPT-small'
        self.tokenizer_name = 'microsoft/DialoGPT-small'
        self.cache_dir = 'new_cached'
        self.block_size = 512
        self.do_train = True
        self.do_eval = True
        self.device = "mps"
        self.evaluate_during_training = False
        self.per_gpu_train_batch_size = 4
        self.per_gpu_eval_batch_size = 4
        self.gradient_accumulation_steps = 1
        self.learning_rate = 5e-5
        self.weight_decay = 0.0
        self.adam_epsilon = 1e-8
        self.max_grad_norm = 1.0
        self.num_train_epochs = 3
        self.max_steps = -1
        self.warmup_steps = 0
        self.logging_steps = 1000
        self.save_steps = 3500
        self.save_total_limit = None
        self.eval_all_checkpoints = False
        self.no_cuda = True
        self.overwrite_output_dir = True
        self.overwrite_cache = True
        self.should_continue = False
        self.seed = 42
        self.local_rank = -1
        self.fp16 = False
        self.fp16_opt_level = 'O1'

args = Args()

### Создаем датасет из ранее подготовленных данных

In [5]:
# Let's look at original dataset
all_rick = pd.read_csv('dialog_talk_agent_clean.csv', delimiter=';')
all_rick.head(10)

Context  \
0  Tell me about your personality   
1       I want to know you better   
2                 Define yourself   
3               Describe yourself   
4          tell me about yourself   
5                   all about you   
6    tell me some stuff about you   
7       talk some stuff about you   
8             talk about yourself   
9                  about yourself   

                                   Text Response  
0    Just think of me as the ace up your sleeve.  
1  I can help you work smarter instead of harder  
2  I can help you work smarter instead of harder  
3  I can help you work smarter instead of harder  
4  I can help you work smarter instead of harder  
5  I can help you work smarter instead of harder  
6  I can help you work smarter instead of harder  
7  I can help you work smarter instead of harder  
8  I can help you work smarter instead of harder  
9  I can help you work smarter instead of harder

Теперь нам нужно создать историю диалога. Будем брать пары ответов из `Context` -> `Text Response`, зацикливая их в один диалог определенной длины

```python
n = 15  # длина диалога
```

In [6]:
contexted = []

n = 15

for i in range(n, len(all_rick['Context']), 2):
  row = []
  prev = i - 1 - n # we additionally substract 1, so row will contain current responce and 7 previous responces  
  for j in range(i, prev, -1):
    if j % 2 == 0:
      row.append(all_rick['Context'][j])
    else:
      row.append(all_rick['Text Response'][j])
  contexted.append(row)  

In [7]:
len(contexted)

746

Cоздадим колонки

In [8]:
columns = ['response', 'context'] 
columns = columns + ['context/'+str(i) for i in range(n-1)]
columns

['response',
 'context',
 'context/0',
 'context/1',
 'context/2',
 'context/3',
 'context/4',
 'context/5',
 'context/6',
 'context/7',
 'context/8',
 'context/9',
 'context/10',
 'context/11',
 'context/12',
 'context/13']

In [9]:
df = pd.DataFrame.from_records(contexted, columns=columns)
df.head(5)

response  \
0      I can help you work smarter instead of harder   
1      I can help you work smarter instead of harder   
2  I'm a relatively new bot, but I'm wise beyond ...   
3  I'm a relatively new bot, but I'm wise beyond ...   
4  I'm a relatively new bot, but I'm wise beyond ...   

                    context  \
0  what is your personality   
1         tell me about you   
2      are you 21 years old   
3           how old are you   
4  how old is your platform   

                                           context/0  \
0      I can help you work smarter instead of harder   
1      I can help you work smarter instead of harder   
2      I can help you work smarter instead of harder   
3  I'm a relatively new bot, but I'm wise beyond ...   
4  I'm a relatively new bot, but I'm wise beyond ...   

                       context/1  \
0  I want to know more about you   
1       what is your personality   
2              tell me about you   
3           are you 21 years old   
4                how old are you   

                                           context/2  \
0      I can help you work smarter instead of harder   
1      I can help you work smarter instead of harder   
2      I can help you work smarter instead of harder   
3      I can help you work smarter instead of harder   
4  I'm a relatively new bot, but I'm wise beyond ...   

                       context/3  \
0                    who are you   
1  I want to know more about you   
2       what is your personality   
3              tell me about you   
4           are you 21 years old   

                                       context/4  \
0  I can help you work smarter instead of harder   
1  I can help you work smarter instead of harder   
2  I can help you work smarter instead of harder   
3  I can help you work smarter instead of harder   
4  I can help you work smarter instead of harder   

                       context/5  \
0            talk about yourself   
1                    who are you   
2  I want to know more about you   
3       what is your personality   
4              tell me about you   

                                       context/6  \
0  I can help you work smarter instead of harder   
1  I can help you work smarter instead of harder   
2  I can help you work smarter instead of harder   
3  I can help you work smarter instead of harder   
4  I can help you work smarter instead of harder   

                       context/7  \
0   tell me some stuff about you   
1            talk about yourself   
2                    who are you   
3  I want to know more about you   
4       what is your personality   

                                       context/8  \
0  I can help you work smarter instead of harder   
1  I can help you work smarter instead of harder   
2  I can help you work smarter instead of harder   
3  I can help you work smarter instead of harder   
4  I can help you work smarter instead of harder   

                       context/9  \
0         tell me about yourself   
1   tell me some stuff about you   
2            talk about yourself   
3                    who are you   
4  I want to know more about you   

                                      context/10  \
0  I can help you work smarter instead of harder   
1  I can help you work smarter instead of harder   
2  I can help you work smarter instead of harder   
3  I can help you work smarter instead of harder   
4  I can help you work smarter instead of harder   

                     context/11  \
0               Define yourself   
1        tell me about yourself   
2  tell me some stuff about you   
3           talk about yourself   
4                   who are you   

                                      context/12  \
0  I can help you work smarter instead of harder   
1  I can help you work smarter instead of harder   
2  I can help you work smarter instead of harder   
3  I can help you work smarter instead of harder   
4  I can help you work smarter instead

Разделим наш датасет на тренировочный и валидационный (90/10)

In [10]:
trn_df, val_df = train_test_split(df, test_size = 0.1)
trn_df.head()

response  \
309                                          Terrific!   
560                                             Howdy.   
42                   Flattery will get you everywhere.   
216  I must have impressed you if you think I'm rea...   
680                     Thanks! The feeling is mutual.   

                    context  \
309             that's nice   
560     lovely day isn't it   
42   you look amazing today   
216            are you real   
680          I like you too   

                                             context/0  \
309                                          Terrific!   
560                                             Howdy.   
42                   Flattery will get you everywhere.   
216  I must have impressed you if you think I'm rea...   
680                     Thanks! The feeling is mutual.   

                     context/1                          context/2  \
309         that's really good                          Terrific!   
560                 a good day                             Howdy.   
42    I think you're beautiful  Flattery will get you everywhere.   
216  I don't think you're fake       Sure! What can I do for you?   
680      but I really like you     Thanks! The feeling is mutual.   

                      context/3                              context/4  \
309                that's great                              Terrific!   
560        just going to say hi                                 Howdy.   
42   you are so beautiful today  Thank you! What a sweet thing to say.   
216         have you been ready           Sure! What can I do for you?   
680               hi I like you         Thanks! The feeling is mutual.   

                context/5                       context/6  \
309           really good                       Terrific!   
560           hello there                          Howdy.   
42                  study   I'm definitely working on it.   
216   are you ready today    Sure! What can I do for you?   
680  of course I like you  Thanks! The feeling is mutual.   

                  context/7                       context/8  \
309         super fantastic                       Terrific!   
560               hey there                          Howdy.   
42                be clever   I'm definitely working on it.   
216       are you ready now                         Always!   
680  I like you you're nice  Thanks! The feeling is mutual.   

                 context/9                      context/10  \
309          this is great                       Terrific!   
560               hi there                          Howdy.   
42           get qualified   I'm definitely working on it.   
216     were you born here            I wish I knew where.   
680  I like that about you  Thanks! The feeling is mutual.   

              context/11                      context/12  \
309    glad to hear that                       Terrific!   
560          I greet you                            Hey!   
42   can you get smarter   I'm definitely working on it.   
216   where are you from            I wish I knew where.   
680     you are so sweet  Thanks! The feeling is mutual.   

                 context/13  
309            good to know  
560                     hey  
42           you must learn  
216  where do you come from  
680      I think I like you

Теперь преобразуем наш набор данных в формат, подходящий для нашей модели. В основном мы будем объединять ответы в одну строку для каждой строки (дополнительно мы добавим специальный токен 'end of string' между ответами, чтобы модель понимала конец каждого ответа в строке).

In [11]:
def construct_conv(row, tokenizer, eos = True):
    flatten = lambda l: [item for sublist in l for item in sublist]
    conv = list(reversed([tokenizer.encode(x) + [tokenizer.eos_token_id] for x in row]))
    conv = flatten(conv)
    return conv

class ConversationDataset(Dataset):
    def __init__(self, tokenizer: PreTrainedTokenizer, args, df, block_size=512):

        block_size = block_size - (tokenizer.model_max_length - tokenizer.max_len_single_sentence)

        directory = args.cache_dir
        cached_features_file = os.path.join(
            directory, args.model_type + "_cached_lm_" + str(block_size)
        )

        if os.path.exists(cached_features_file) and not args.overwrite_cache:
            logger.info("Loading features from cached file %s", cached_features_file)
            with open(cached_features_file, "rb") as handle:
                self.examples = pickle.load(handle)
        else:
            logger.info("Creating features from dataset file at %s", directory)

            self.examples = []
            for _, row in df.iterrows():
                conv = construct_conv(row, tokenizer)
                self.examples.append(conv)

            logger.info("Saving features into cached file %s", cached_features_file)
            with open(cached_features_file, "wb") as handle:
                pickle.dump(self.examples, handle, protocol=pickle.HIGHEST_PROTOCOL)

    def __len__(self):
        return len(self.examples)

    def __getitem__(self, item):
        return torch.tensor(self.examples[item], dtype=torch.long)

In [12]:
# Cacheing and storing of data/checkpoints

def load_and_cache_examples(args, tokenizer, df_trn, df_val, evaluate=False):
    return ConversationDataset(tokenizer, args, df_val if evaluate else df_trn)


def set_seed(args):
    random.seed(args.seed)
    np.random.seed(args.seed)
    torch.manual_seed(args.seed)
    if args.n_gpu > 0:
        torch.cuda.manual_seed_all(args.seed)


def _sorted_checkpoints(args, checkpoint_prefix="checkpoint", use_mtime=False) -> List[str]:
    ordering_and_checkpoint_path = []

    glob_checkpoints = glob.glob(os.path.join(args.output_dir, "{}-*".format(checkpoint_prefix)))

    for path in glob_checkpoints:
        if use_mtime:
            ordering_and_checkpoint_path.append((os.path.getmtime(path), path))
        else:
            regex_match = re.match(".*{}-([0-9]+)".format(checkpoint_prefix), path)
            if regex_match and regex_match.groups():
                ordering_and_checkpoint_path.append((int(regex_match.groups()[0]), path))

    checkpoints_sorted = sorted(ordering_and_checkpoint_path)
    checkpoints_sorted = [checkpoint[1] for checkpoint in checkpoints_sorted]
    return checkpoints_sorted


def _rotate_checkpoints(args, checkpoint_prefix="checkpoint", use_mtime=False) -> None:
    if not args.save_total_limit:
        return
    if args.save_total_limit <= 0:
        return

    # Check if we should delete older checkpoint(s)
    checkpoints_sorted = _sorted_checkpoints(args, checkpoint_prefix, use_mtime)
    if len(checkpoints_sorted) <= args.save_total_limit:
        return

    number_of_checkpoints_to_delete = max(0, len(checkpoints_sorted) - args.save_total_limit)
    checkpoints_to_be_deleted = checkpoints_sorted[:number_of_checkpoints_to_delete]
    for checkpoint in checkpoints_to_be_deleted:
        logger.info("Deleting older checkpoint [{}] due to args.save_total_limit".format(checkpoint))
        shutil.rmtree(checkpoint)

Функция треннировочного и валидационного цикла

```python
def train(...):
    ...


def evaluate(...):
    ...
```

In [13]:
def train(args, train_dataset, model: PreTrainedModel, tokenizer: PreTrainedTokenizer) -> Tuple[int, float]:
    """ Train the model """
    if args.local_rank in [-1, 0]:
        tb_writer = SummaryWriter()

    args.train_batch_size = args.per_gpu_train_batch_size * max(1, args.n_gpu)

    def collate(examples: List[torch.Tensor]):
        if tokenizer._pad_token is None:
            return pad_sequence(examples, batch_first=True)
        return pad_sequence(examples, batch_first=True, padding_value=tokenizer.pad_token_id)

    train_sampler = RandomSampler(train_dataset) if args.local_rank == -1 else DistributedSampler(train_dataset)
    train_dataloader = DataLoader(
        train_dataset, sampler=train_sampler, batch_size=args.train_batch_size, collate_fn=collate, drop_last = True
    )

    if args.max_steps > 0:
        t_total = args.max_steps
        args.num_train_epochs = args.max_steps // (len(train_dataloader) // args.gradient_accumulation_steps) + 1
    else:
        t_total = len(train_dataloader) // args.gradient_accumulation_steps * args.num_train_epochs

    model = model.module if hasattr(model, "module") else model  # Take care of distributed/parallel training
    model.resize_token_embeddings(len(tokenizer))
    # add_special_tokens_(model, tokenizer)


    # Prepare optimizer and schedule (linear warmup and decay)
    no_decay = ["bias", "LayerNorm.weight"]
    optimizer_grouped_parameters = [
        {
            "params": [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)],
            "weight_decay": args.weight_decay,
        },
        {"params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], "weight_decay": 0.0},
    ]
    optimizer = AdamW(optimizer_grouped_parameters, lr=args.learning_rate, eps=args.adam_epsilon)
    scheduler = get_linear_schedule_with_warmup(
        optimizer, num_warmup_steps=args.warmup_steps, num_training_steps=t_total
    )

    # Check if saved optimizer or scheduler states exist
    if (
        args.model_name_or_path
        and os.path.isfile(os.path.join(args.model_name_or_path, "optimizer.pt"))
        and os.path.isfile(os.path.join(args.model_name_or_path, "scheduler.pt"))
    ):
        # Load in optimizer and scheduler states
        optimizer.load_state_dict(torch.load(os.path.join(args.model_name_or_path, "optimizer.pt")))
        scheduler.load_state_dict(torch.load(os.path.join(args.model_name_or_path, "scheduler.pt")))

    if args.fp16:
        try:
            from apex import amp
        except ImportError:
            raise ImportError("Please install apex from https://www.github.com/nvidia/apex to use fp16 training.")
        model, optimizer = amp.initialize(model, optimizer, opt_level=args.fp16_opt_level)

    # multi-gpu training (should be after apex fp16 initialization)
    if args.n_gpu > 1:
        model = torch.nn.DataParallel(model)

    # Distributed training (should be after apex fp16 initialization)
    if args.local_rank != -1:
        model = torch.nn.parallel.DistributedDataParallel(
            model, device_ids=[args.local_rank], output_device=args.local_rank, find_unused_parameters=True
        )

    # Train!
    logger.info("***** Running training *****")
    logger.info("  Num examples = %d", len(train_dataset))
    logger.info("  Num Epochs = %d", args.num_train_epochs)
    logger.info("  Instantaneous batch size per GPU = %d", args.per_gpu_train_batch_size)
    logger.info(
        "  Total train batch size (w. parallel, distributed & accumulation) = %d",
        args.train_batch_size
        * args.gradient_accumulation_steps
        * (torch.distributed.get_world_size() if args.local_rank != -1 else 1),
    )
    logger.info("  Gradient Accumulation steps = %d", args.gradient_accumulation_steps)
    logger.info("  Total optimization steps = %d", t_total)

    global_step = 0
    epochs_trained = 0
    steps_trained_in_current_epoch = 0
    # Check if continuing training from a checkpoint
    if args.model_name_or_path and os.path.exists(args.model_name_or_path):
        try:
            # set global_step to gobal_step of last saved checkpoint from model path
            checkpoint_suffix = args.model_name_or_path.split("-")[-1].split("/")[0]
            global_step = int(checkpoint_suffix)
            epochs_trained = global_step // (len(train_dataloader) // args.gradient_accumulation_steps)
            steps_trained_in_current_epoch = global_step % (len(train_dataloader) // args.gradient_accumulation_steps)

            logger.info("  Continuing training from checkpoint, will skip to saved global_step")
            logger.info("  Continuing training from epoch %d", epochs_trained)
            logger.info("  Continuing training from global step %d", global_step)
            logger.info("  Will skip the first %d steps in the first epoch", steps_trained_in_current_epoch)
        except ValueError:
            logger.info("  Starting fine-tuning.")

    tr_loss, logging_loss = 0.0, 0.0

    model.zero_grad()
    train_iterator = trange(
        epochs_trained, int(args.num_train_epochs), desc="Epoch", disable=args.local_rank not in [-1, 0]
    )
    set_seed(args)  # Added here for reproducibility
    for _ in train_iterator:
        epoch_iterator = tqdm(train_dataloader, desc="Iteration", disable=args.local_rank not in [-1, 0])
        for step, batch in enumerate(epoch_iterator):

            # Skip past any already trained steps if resuming training
            if steps_trained_in_current_epoch > 0:
                steps_trained_in_current_epoch -= 1
                continue

            inputs, labels = (batch, batch)
            if inputs.shape[1] > 1024: continue
            inputs = inputs.to(args.device)
            labels = labels.to(args.device)
            model.train()
            outputs = model(inputs, labels=labels)
            loss = outputs[0]  # model outputs are always tuple in transformers (see doc)

            if args.n_gpu > 1:
                loss = loss.mean()  # mean() to average on multi-gpu parallel training
            if args.gradient_accumulation_steps > 1:
                loss = loss / args.gradient_accumulation_steps

            if args.fp16:
                with amp.scale_loss(loss, optimizer) as scaled_loss:
                    scaled_loss.backward()
            else:
                loss.backward()

            tr_loss += loss.item()
            if (step + 1) % args.gradient_accumulation_steps == 0:
                if args.fp16:
                    torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), args.max_grad_norm)
                else:
                    torch.nn.utils.clip_grad_norm_(model.parameters(), args.max_grad_norm)
                optimizer.step()
                scheduler.step()  # Update learning rate schedule
                model.zero_grad()
                global_step += 1

                if args.local_rank in [-1, 0] and args.logging_steps > 0 and global_step % args.logging_steps == 0:
                    # Log metrics
                    if (
                        args.local_rank == -1 and args.evaluate_during_training
                    ):  # Only evaluate when single GPU otherwise metrics may not average well
                        results = evaluate(args, model, tokenizer)
                        for key, value in results.items():
                            tb_writer.add_scalar("eval_{}".format(key), value, global_step)
                    tb_writer.add_scalar("lr", scheduler.get_lr()[0], global_step)
                    tb_writer.add_scalar("loss", (tr_loss - logging_loss) / args.logging_steps, global_step)
                    logging_loss = tr_loss

                if args.local_rank in [-1, 0] and args.save_steps > 0 and global_step % args.save_steps == 0:
                    checkpoint_prefix = "checkpoint"
                    # Save model checkpoint
                    output_dir = os.path.join(args.output_dir, "{}-{}".format(checkpoint_prefix, global_step))
                    os.makedirs(output_dir, exist_ok=True)
                    model_to_save = (
                        model.module if hasattr(model, "module") else model
                    )  # Take care of distributed/parallel training
                    model_to_save.save_pretrained(output_dir)
                    tokenizer.save_pretrained(output_dir)

                    torch.save(args, os.path.join(output_dir, "training_args.bin"))
                    logger.info("Saving model checkpoint to %s", output_dir)

                    _rotate_checkpoints(args, checkpoint_prefix)

                    torch.save(optimizer.state_dict(), os.path.join(output_dir, "optimizer.pt"))
                    torch.save(scheduler.state_dict(), os.path.join(output_dir, "scheduler.pt"))
                    logger.info("Saving optimizer and scheduler states to %s", output_dir)

            if args.max_steps > 0 and global_step > args.max_steps:
                epoch_iterator.close()
                break
        if args.max_steps > 0 and global_step > args.max_steps:
            train_iterator.close()
            break

    if args.local_rank in [-1, 0]:
        tb_writer.close()

    return global_step, tr_loss / global_step

# Evaluation of some model

def evaluate(args, model: PreTrainedModel, tokenizer: PreTrainedTokenizer, df_trn, df_val, prefix="") -> Dict:
    # Loop to handle MNLI double evaluation (matched, mis-matched)
    eval_output_dir = args.output_dir

    eval_dataset = load_and_cache_examples(args, tokenizer, df_trn, df_val, evaluate=True)
    os.makedirs(eval_output_dir, exist_ok=True)
    args.eval_batch_size = args.per_gpu_eval_batch_size * max(1, args.n_gpu)
    # Note that DistributedSampler samples randomly

    def collate(examples: List[torch.Tensor]):
        if tokenizer._pad_token is None:
            return pad_sequence(examples, batch_first=True)
        return pad_sequence(examples, batch_first=True, padding_value=tokenizer.pad_token_id)

    eval_sampler = SequentialSampler(eval_dataset)
    eval_dataloader = DataLoader(
        eval_dataset, sampler=eval_sampler, batch_size=args.eval_batch_size, collate_fn=collate, drop_last = True
    )

    # multi-gpu evaluate
    if args.n_gpu > 1:
        model = torch.nn.DataParallel(model)

    # Eval!
    logger.info("***** Running evaluation {} *****".format(prefix))
    logger.info("  Num examples = %d", len(eval_dataset))
    logger.info("  Batch size = %d", args.eval_batch_size)
    eval_loss = 0.0
    nb_eval_steps = 0
    model.eval()

    for batch in tqdm(eval_dataloader, desc="Evaluating"):
        inputs, labels = (batch, batch)
        inputs = inputs.to(args.device)
        labels = labels.to(args.device)

        with torch.no_grad():
            outputs = model(inputs, labels=labels)
            lm_loss = outputs[0]
            eval_loss += lm_loss.mean().item()
        nb_eval_steps += 1

    eval_loss = eval_loss / nb_eval_steps
    perplexity = torch.exp(torch.tensor(eval_loss))

    result = {"perplexity": perplexity}

    output_eval_file = os.path.join(eval_output_dir, prefix, "eval_results.txt")
    with open(output_eval_file, "w") as writer:
        logger.info("***** Eval results {} *****".format(prefix))
        for key in sorted(result.keys()):
            logger.info("  %s = %s", key, str(result[key]))
            writer.write("%s = %s\n" % (key, str(result[key])))

    return result

Основная функция запуска цикла обучения

In [14]:
# Main runner

def main(df_trn, df_val):
    args = Args()
    
    if args.should_continue:
        sorted_checkpoints = _sorted_checkpoints(args)
        if len(sorted_checkpoints) == 0:
            raise ValueError("Used --should_continue but no checkpoint was found in --output_dir.")
        else:
            args.model_name_or_path = sorted_checkpoints[-1]

    if (
        os.path.exists(args.output_dir)
        and os.listdir(args.output_dir)
        and args.do_train
        and not args.overwrite_output_dir
        and not args.should_continue
    ):
        raise ValueError(
            "Output directory ({}) already exists and is not empty. Use --overwrite_output_dir to overcome.".format(
                args.output_dir
            )
        )

    # Setup CUDA, GPU & distributed training
    device = torch.device("mps")
    args.n_gpu = 0
    args.device = device

    # Setup logging
    logging.basicConfig(
        format="%(asctime)s - %(levelname)s - %(name)s -   %(message)s",
        datefmt="%m/%d/%Y %H:%M:%S",
        level=logging.INFO if args.local_rank in [-1, 0] else logging.WARN,
    )
    logger.warning(
        "Process rank: %s, device: %s, n_gpu: %s, distributed training: %s, 16-bits training: %s",
        args.local_rank,
        device,
        args.n_gpu,
        bool(args.local_rank != -1),
        args.fp16,
    )

    # Set seed
    set_seed(args)

    config = AutoConfig.from_pretrained(args.config_name, cache_dir=args.cache_dir)
    tokenizer = AutoTokenizer.from_pretrained(args.tokenizer_name, cache_dir=args.cache_dir)
    model = AutoModelWithLMHead.from_pretrained(
        args.model_name_or_path,
        from_tf=False,
        config=config,
        cache_dir=args.cache_dir,
    )
    model.to(args.device)
    
    logger.info("Training/evaluation parameters %s", args)

    # Training
    if args.do_train:
        train_dataset = load_and_cache_examples(args, tokenizer, df_trn, df_val, evaluate=False)

        global_step, tr_loss = train(args, train_dataset, model, tokenizer)
        logger.info(" global_step = %s, average loss = %s", global_step, tr_loss)

    # Saving best-practices: if you use save_pretrained for the model and tokenizer, you can reload them using from_pretrained()
    if args.do_train:
        # Create output directory if needed
        os.makedirs(args.output_dir, exist_ok=True)

        logger.info("Saving model checkpoint to %s", args.output_dir)
        # Save a trained model, configuration and tokenizer using `save_pretrained()`.
        # They can then be reloaded using `from_pretrained()`
        model_to_save = (
            model.module if hasattr(model, "module") else model
        )  # Take care of distributed/parallel training
        model_to_save.save_pretrained(args.output_dir)
        tokenizer.save_pretrained(args.output_dir)

        # Good practice: save your training arguments together with the trained model
        torch.save(args, os.path.join(args.output_dir, "training_args.bin"))

        # Load a trained model and vocabulary that you have fine-tuned
        model = AutoModelWithLMHead.from_pretrained(args.output_dir)
        tokenizer = AutoTokenizer.from_pretrained(args.output_dir)
        model.to(args.device)

    # Evaluation
    results = {}
    if args.do_eval and args.local_rank in [-1, 0]:
        checkpoints = [args.output_dir]
        if args.eval_all_checkpoints:
            checkpoints = list(
                os.path.dirname(c) for c in sorted(glob.glob(args.output_dir + "/**/" + WEIGHTS_NAME, recursive=True))
            )
            logging.getLogger("transformers.modeling_utils").setLevel(logging.WARN)  # Reduce logging
        logger.info("Evaluate the following checkpoints: %s", checkpoints)
        for checkpoint in checkpoints:
            global_step = checkpoint.split("-")[-1] if len(checkpoints) > 1 else ""
            prefix = checkpoint.split("/")[-1] if checkpoint.find("checkpoint") != -1 else ""

            model = AutoModelWithLMHead.from_pretrained(checkpoint)
            model.to(args.device)
            result = evaluate(args, model, tokenizer, df_trn, df_val, prefix=prefix)
            result = dict((k + "_{}".format(global_step), v) for k, v in result.items())
            results.update(result)

    return results

Начнем fine-tuning модели

In [15]:
main(trn_df, val_df)

11/24/2022 00:07:10 - WARNING - __main__ -   Process rank: -1, device: mps, n_gpu: 0, distributed training: False, 16-bits training: False
/opt/homebrew/Caskroom/miniforge/base/envs/torch_lightning/lib/python3.9/site-packages/transformers/models/auto/modeling_auto.py:1132: FutureWarning: The class `AutoModelWithLMHead` is deprecated and will be removed in a future version. Please use `AutoModelForCausalLM` for causal language models, `AutoModelForMaskedLM` for masked language models and `AutoModelForSeq2SeqLM` for encoder-decoder models.
  warnings.warn(
11/24/2022 00:07:15 - INFO - __main__ -   Training/evaluation parameters <__main__.Args object at 0x2c490f100>
11/24/2022 00:07:15 - INFO - __main__ -   Creating features from dataset file at new_cached
11/24/2022 00:07:16 - INFO - __main__ -   Saving features into cached file new_cached/gpt2_cached_lm_512
/opt/homebrew/Caskroom/miniforge/base/envs/torch_lightning/lib/python3.9/site-packages/transformers/optimization.py:306: FutureWarn

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Epoch:   0%|          | 0/3 [00:00<?, ?it/s]

Iteration:   0%|          | 0/167 [00:00<?, ?it/s]

Iteration:   0%|          | 0/167 [00:00<?, ?it/s]

Iteration:   0%|          | 0/167 [00:00<?, ?it/s]

11/24/2022 00:35:53 - INFO - __main__ -    global_step = 501, average loss = 0.8412926584303736
11/24/2022 00:35:53 - INFO - __main__ -   Saving model checkpoint to new_output-small
11/24/2022 00:36:07 - INFO - __main__ -   Evaluate the following checkpoints: ['new_output-small']
11/24/2022 00:36:15 - INFO - __main__ -   Creating features from dataset file at new_cached
11/24/2022 00:36:15 - INFO - __main__ -   Saving features into cached file new_cached/gpt2_cached_lm_512
11/24/2022 00:36:15 - INFO - __main__ -   ***** Running evaluation  *****
11/24/2022 00:36:15 - INFO - __main__ -     Num examples = 75
11/24/2022 00:36:15 - INFO - __main__ -     Batch size = 4


Evaluating:   0%|          | 0/18 [00:00<?, ?it/s]

11/24/2022 00:36:36 - INFO - __main__ -   ***** Eval results  *****
11/24/2022 00:36:36 - INFO - __main__ -     perplexity = tensor(1.5382)


{'perplexity_': tensor(1.5382)}

Метрика обучения модели - `perplexity` <br>
Подробнее: https://huggingface.co/docs/transformers/perplexity

Наш `perplexity score` = 1.5382

### Загрузим модель и сгенерируем ответы

Здесь есть важное условие, от которого мы будем отталкиваться в дальнейшем

Для DialoGPT характерна история диалога как контекст чата, для которого будет сгенерирован ответ. <br>
В историю диалога входят как сообщения от пользователя, так и ответы от модели

В ином случае, мы не передаем историю диалога модели и каждый раз идем с *холодного старта*

In [16]:
tokenizer = AutoTokenizer.from_pretrained('microsoft/DialoGPT-small')
model = AutoModelWithLMHead.from_pretrained('new_output-small')


def chat(text, chat_history_ids):
    # encode the new user input, add the eos_token and return a tensor in Pytorch
    new_user_input_ids = tokenizer.encode(text + tokenizer.eos_token, return_tensors='pt')
    # print(new_user_input_ids)

    # append the new user input tokens to the chat history
    bot_input_ids = torch.cat([chat_history_ids, new_user_input_ids], dim=-1) if chat_history_ids is not None else new_user_input_ids

    # generated a response while limiting the total chat history to 1000 tokens, 
    chat_history_ids = model.generate(
        bot_input_ids, max_length=200,
        pad_token_id=tokenizer.eos_token_id,  
        no_repeat_ngram_size=3,       
        do_sample=True, 
        top_k=100, 
        top_p=0.7,
        temperature = 0.8
    )
    
    # pretty print last ouput tokens from bot
    print("DialogBot: {}".format(tokenizer.decode(chat_history_ids[:, bot_input_ids.shape[-1]:][0], skip_special_tokens=True)))

    return chat_history_ids

/opt/homebrew/Caskroom/miniforge/base/envs/torch_lightning/lib/python3.9/site-packages/transformers/models/auto/modeling_auto.py:1132: FutureWarning: The class `AutoModelWithLMHead` is deprecated and will be removed in a future version. Please use `AutoModelForCausalLM` for causal language models, `AutoModelForMaskedLM` for masked language models and `AutoModelForSeq2SeqLM` for encoder-decoder models.
  warnings.warn(


In [17]:
chat_history_ids = None

chat_history_ids = chat("Tell me about your personality", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: I'm flattered by your personality.


In [18]:
chat_history_ids = chat("Who are you", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: Flattery will get you everywhere.


In [19]:
chat_history_ids = chat("Answer the question", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: I want to know more about you


In [20]:
chat_history_ids = chat("Are you 21 years old?", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: I am


In [21]:
chat_history_ids = chat("Are you 21 old?", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: are you 21


In [22]:
chat_history_ids = chat("I find you annoying", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: Are we best friends ever?


In [23]:
chat_history_ids = chat("Yes, we are", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: How old are you?


In [24]:
chat_history_ids = chat("I'm 20", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: How's your day been?


In [25]:
chat_history_ids = chat("Fine, thank you!", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: !!!!!!?!!,!!.!!


In [26]:
chat_history_ids = chat("Okay, tell me about you", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: !?,!,,!??!,?!?


In [27]:
chat_history_ids = None

chat_history_ids = chat("Okay, tell me about you", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: I'm always right here.


In [28]:
chat_history_ids = chat("Are you bot?", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: I am always right where I left you


In [29]:
chat_history_ids = chat("Just answer me, are you bot?", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: are you a bot


In [30]:
chat_history_ids = chat("No, im not", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: You're very special to me.


In [31]:
chat_history_ids = chat("That's a lot", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: That is very true.


In [32]:
chat_history_ids = chat("true", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: That would be true. But I'm not special


In [33]:
chat_history_ids = chat("Can you write a code for me?", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: !!!!!!,!!.!!?!!


In [34]:
chat_history_ids = None

chat_history_ids = chat("give me the answer", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: I'm not programmed for that exact question. Try asking another way?


In [35]:
chat_history_ids = chat("Can you give me the answer?", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: I am programmed for this exact question


In [51]:
chat_history_ids = None

chat_history_ids = chat("who are you", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: You know I'm not really sure. I'm just a baby bot 🤖.


In [52]:
chat_history_ids = None

chat_history_ids = chat("about yourself", chat_history_ids)

A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


DialogBot: Thanks! The feeling is mutual.


### Загрузим нашу модель на Hugging Face Hub

In [53]:
from huggingface_hub import notebook_login

notebook_login()

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Login successful
Your token has been saved to /Users/nikitadavidcuk/.huggingface/token
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Authenticated through git-credential store but this isn't the helper defined on your machine.
You might have to re-authenticate when pushing to the Hugging Face Hub. Run the following command in your terminal in case you want to set this credential helper as the de

In [54]:
model.push_to_hub("Ar4ikov/DialogAgentGPT2")
tokenizer.push_to_hub("Ar4ikov/DialogAgentGPT2")

CommitInfo(commit_url='https://huggingface.co/Ar4ikov/DialogAgentGPT2/commit/8f0f26c15dc4e461d509b4b65ee95ce55d3dfec3', commit_message='Upload tokenizer', commit_description='', oid='8f0f26c15dc4e461d509b4b65ee95ce55d3dfec3', pr_url=None, pr_revision=None, pr_num=None)

Теперь наша зафайнтюненная модель и токенайзер доступны для всех по адресу <br>
> https://huggingface.co/Ar4ikov/DialogAgentGPT2

В дальнейшем их можно использовать в пайплайне для создания чат-бота

(Картинку в иконке репозитория сменю потом)

![img](https://i.imgur.com/bGXIWqm.png)